src/sage/rings/species.py

```diff
@@ -0,0 +1,1250 @@
+
+                sage: P = PolynomialSpecies(["X"])
+                sage: M = P._indices
+                sage: X = M(SymmetricGroup(1))
+                sage: E2 = M(SymmetricGroup(2))
+                sage: E2(X)
+                E_2
+                sage: X(E2)
+                E_2
+                sage: E2(E2)
+                P_4
+            """
+            if len(args) != self.parent()._k:
+                raise ValueError("number of args must match arity of self")
+            if not all(isinstance(arg, MolecularSpecies.Element) for arg in args):
+                raise ValueError("all args must be molecular species")
+            if len(set(arg.parent()._k for arg in args)) > 1:
+                raise ValueError("all args must have same arity")
+            if len(set(arg._mc for arg in args)) > 1:
+                raise ValueError("all args must have same multicardinality")
+
+            gens = []
+
+            # TODO: What happens if in F(G), G has a constant part? E(1+X)?
+            Mlist = [None for _ in range(self._group.degree())]
+            for i, v in enumerate(self._dompart):
+                for k in v:
+                    Mlist[k - 1] = args[i]
+            starts = list(accumulate([M._group.degree() for M in Mlist], initial=0))
+
+            # gens from self
+            for gen in self._group.gens():
+                newgen = []
+                for cyc in gen.cycle_tuples():
+                    for k in range(1, Mlist[cyc[0] - 1]._group.degree() + 1):
+                        newgen.append(tuple(k + starts[i - 1] for i in cyc))
+                gens.append(newgen)
+
+            # gens from M_i and dompart
+            dpart = {i: [] for i in range(1, args[0].parent()._k + 1)}
+            for start, M in zip(starts, Mlist):
+                for i, v in enumerate(M._dompart, 1):
+                    dpart[i].extend([start + k for k in v])
+                for gen in M._group.gens():
+                    gens.append([tuple(start + k for k in cyc) for cyc in gen.cycle_tuples()])
+
+            return args[0].parent()(PermutationGroup(gens, domain=range(1, starts[-1] + 1)), dpart)
+
+
+class PolynomialSpecies(CombinatorialFreeModule):
+    def __classcall__(cls, k, singleton_names=None, base_ring=ZZ):
+        r"""
+        Normalize the arguments.
+
+        TESTS::
+
+            sage: P1 = PolynomialSpecies(1, "X", ZZ)
+            sage: P2 = PolynomialSpecies(1, base_ring=ZZ)
+            sage: P3 = PolynomialSpecies(["X", "Y"], base_ring=ZZ)
+            sage: P4 = PolynomialSpecies(2, ["X", "Y"])
+            sage: P1 is P2
+            False
+            sage: P3 is P4
+            True
+
+        .. TODO::
+
+            Reconsider the order of the arguments.
+            ``singleton_names`` are not generators, and may even be
+            omitted, and ``base_ring`` will usually be ``ZZ``, but
+            maybe it would be nice to allow ``P.<X,Y,Z> =
+            PolynomialSpecies(ZZ)`` anyway.
```

Comment:
In line with the comment above, we really want ``PolynomialSpecies(base_ring, names)``.